### PR TITLE
Allow for custom sharding specs to register their own ops.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76477
* __->__ #76360

Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

### CODE REVIEW

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 1]
```diff
@@ -870,6 +885,15 @@ def __truediv__(self, other):
     def __rtruediv__(self, other):
         return handle_torch_function(torch.Tensor.__rdiv__, (self, other), self, other)
 
+    def tanh(self):
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/api.py:888]: why we need to special register these three ops in the `ShardedTensor` class?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+)
+from torch.distributed._shard._utils import narrow_tensor
+import torch.distributed as dist
+from typing import List, Union, TYPE_CHECKING
+from ._internals import (
+    get_chunked_dim_size,
+    get_split_size,
+)
+
+from . import ShardingSpec
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:17]: nit: `from .api import ShardingSpec` to avoid circular imports?

File: torch/distributed/_shard/sharded_tensor/_ops/math_ops.py [hunk 3]
```diff
@@ -4,96 +4,80 @@
...
 
-            return ShardedTensor._init_from_local_tensor(
-                res,
-                rhs.sharding_spec(),
-                rhs.size(),  # type: ignore[arg-type]
-                process_group=pg)
+    elif isinstance(rhs, ReplicatedTensor):
+        assert isinstance(lhs, ShardedTensor)
+        st_size = lhs.size()
+        st_meta = lhs.local_shards()[0].metadata
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:42]: (no need to do it in this PR) Do you think for spec agnostic sharded tensor ops, we should consider the case of multiple local shards?

File: torch/distributed/_shard/sharding_spec/api.py [hunk 4]
```diff
@@ -92,174 +88,60 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
...
+        Args:
+            sharding_spec(type): The ShardingSpec for which we need to add this custom op.
+            op(Callable): The op to override (ex: torch.bmm)
+            func(Callable): The custom implementation for ``op``
+        """
+        from inspect import signature
+        if len(signature(func).parameters) != 3:
+            raise TypeError(
+                f'Custom sharded op function expects signature: '
+                f'(types, args, kwargs), but received '
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/api.py:105]: for custom_spec_op, we don't pass process_group, are we assuming that we would use the sharded tensor associated process group for communication in the op implementation? what if two parameters in the module have different process group, should we check that they are in the same pg before doing the computation?

File: torch/distributed/_shard/sharding_spec/api.py [hunk 5]
```diff
@@ -92,174 +88,60 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
...
+                f'signature: {signature(func)}')
+
+        class_name = sharding_spec.__qualname__
+        if class_name not in ShardingSpec.custom_ops:
+            ShardingSpec.custom_ops[class_name] = {}
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/api.py:110]: why we need a class name as a key here, is it here to store all sharding spec op table? If so, do we have a inheritance rule that if a subclass of a ShardingSpec op can't be find, search it in the parent class instead?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 6]
```diff
@@ -39,7 +39,7 @@
 _sharded_tensor_map: Dict[int, 'weakref.ReferenceType[ShardedTensor]'] = {}
 
 # Custom sharded ops
-_SHARDED_OPS: Dict[str, Callable] = {}
+_SHARDED_OPS: Dict[Callable, Callable] = {}
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/api.py:42]: should this `_SHRDED_OPS` be removed, and we consolidate the ops into the base `ShardingSpec` instead?

**Review by wanchaol — COMMENTED** [review]
Looks great! Some questions and suggestions inlined.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 6 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/sharded_tensor/api.py:46]: See comment above, it doesn't seem right to have ops registered on an interface class.

File: torch/distributed/_shard/sharded_tensor/_ops/math_ops.py (hunk 3 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:42]: Indeed, some ops have very complicated behavior under the circumstances of multiple shards, which we need to define the expected behavior first. I guess for simple ops like element-wise ops, we already implemented in a spec agnostic way.

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 7]
```diff
@@ -870,6 +890,15 @@ def __truediv__(self, other):
...
         return handle_torch_function(torch.Tensor.__rdiv__, (self, other), self, other)
 
+    def tanh(self):
+        return handle_torch_function(torch.Tensor.tanh, (self,), self)
+
+    def __getitem__(self, key):
+        return handle_torch_function(torch.Tensor.__getitem__, (self, key), self, key)
+
+    def transpose(self, dim0, dim1):
+        return handle_torch_function(torch.Tensor.transpose, (self, dim0, dim1), self, dim0, dim1)
```
Comment (fduwjj) [inline torch/distributed/_shard/sharded_tensor/api.py:900]: We need meta change for this op right? Btw, transpose has been added in https://github.com/pytorch/pytorch/pull/75374.

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 8]
```diff
@@ -870,6 +890,15 @@ def __truediv__(self, other):
     def __rtruediv__(self, other):
         return handle_torch_function(torch.Tensor.__rdiv__, (self, other), self, other)
 
+    def tanh(self):
+        return handle_torch_function(torch.Tensor.tanh, (self,), self)
```
Comment (fduwjj) [inline torch/distributed/_shard/sharded_tensor/api.py:894]: Nit: can we add this as a new op in the element-wise op?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/math_ops.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            if lhs.size() == rhs.size() and lhs_spec.dim == rhs_spec.dim:
+                # perform local element-wise math op
+                res = op(lhs.local_tensor(), rhs.local_tensor())
+                return ShardedTensor._init_from_local_tensor(
+                    res,
+                    lhs_spec,
+                    lhs.size(),  # type: ignore[arg-type]
+                    process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
```
Comment (fduwjj) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/math_ops.py:40]: Nit: we also don't support when sharding dim is different, right?

File: torch/distributed/_shard/sharded_tensor/_ops/__init__.py [hunk 10]
```diff
@@ -2,7 +2,10 @@
...
-from .embedding import sharded_embedding
-from .embedding_bag import sharded_embedding_bag
 from .init import kaiming_uniform_, normal_, uniform_, constant_
-from .linear import sharded_linear
+
+# Import all ChunkShardingSpec ops
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.linear import sharded_linear
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.embedding import sharded_embedding
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.embedding_bag import sharded_embedding_bag
+import torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.math_ops
```
Comment (fduwjj) [inline torch/distributed/_shard/sharded_tensor/_ops/__init__.py:11]: Curious, why not put this into `torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/__init__.py` first and we then import that file?

**Review by fduwjj — COMMENTED** [review]
Thanks for separating between sharding spec specific ops and general ops. https://github.com/pytorch/pytorch/pull/75374 has been landed and can you kindly rebase on top of it? Thanks!

**Review by wanchaol — APPROVED** [review]
lgtm!

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/math_ops.py (hunk 9 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/math_ops.py:40]: Broadcasting probably covers the sharding dim mismatch case too since in that case you probably need to do broadcasting across ranks.
